src/gsy_e/models/strategy/energy_parameters/heat_pump.py

```diff
@@ -1,48 +1,171 @@
+        return self.state.get_min_energy_demand_kWh(time_slot)
+
+    def get_max_energy_demand_kWh(self, time_slot: DateTime) -> float:
+        """Get energy that is needed to heat up the storage to temp_max."""
+        return self.state.get_max_energy_demand_kWh(time_slot)
+
+    def _next_time_slot(self, current_time_slot: DateTime) -> DateTime:
+        return current_time_slot + self._slot_length
+
+    def _temp_diff_to_energy(self, diff_temp_K: float) -> float:
+        return diff_temp_K * self._Q_specific
+
+    def _energy_to_temp_diff(self, energy_kWh: float) -> float:
+        return energy_kWh / self._Q_specific
+
+    def _rotate_profiles(self, current_time_slot: Optional[DateTime] = None):
+        self._consumption_kWh.read_or_rotate_profiles()
+        self._ext_temp_C.read_or_rotate_profiles()
+
+        self.state.delete_past_state_values(current_time_slot)
+
+    def _calc_energy_to_buy_maximum(self, time_slot: DateTime) -> float:
+        max_energy_consumption = self._temp_diff_to_energy(
+            self._max_temp_C -
+            self.state.get_storage_temp_C(time_slot) +
+            self.state.get_temp_decrease_K(time_slot)) / self._get_cop(time_slot)
+
+        assert max_energy_consumption > -FLOATING_POINT_TOLERANCE
+        return min(self._max_energy_consumption_kWh, max_energy_consumption)
+
+    def _calc_energy_to_buy_minimum(self, time_slot: DateTime) -> float:
+        max_temp_decrease_allowed = (
+                self.state.get_storage_temp_C(time_slot) - self._min_temp_C)
+        temp_diff = self.state.get_temp_decrease_K(time_slot) - max_temp_decrease_allowed
+        if temp_diff < -FLOATING_POINT_TOLERANCE:
+            return 0
+        min_energy_consumption = self._temp_diff_to_energy(temp_diff) / self._get_cop(time_slot)
+        return min(self._max_energy_consumption_kWh, min_energy_consumption)
+
+    def _calc_temp_decrease_K(self, time_slot: DateTime) -> float:
+        return self._energy_to_temp_diff(self._calc_Q_consumption(
+            time_slot, self._consumption_kWh.profile[time_slot]))
+
+    def _calc_temp_increase_K(self, time_slot: DateTime, energy_kWh: float) -> float:
+        return self._energy_to_temp_diff(self._calc_Q_consumption(time_slot, energy_kWh))
+
+    def _populate_state(self, time_slot: DateTime):
+        # order matters here
+        self.state.set_temp_decrease_K(
+            time_slot, self._calc_temp_decrease_K(time_slot))
+
+        self.state.update_storage_temp(time_slot)
+
+        self._calc_energy_demand(time_slot)
+
+    def _calc_energy_demand(self, time_slot: DateTime):
+        self.state.set_min_energy_demand_kWh(
+            time_slot, self._calc_energy_to_buy_minimum(time_slot))
+        self.state.set_max_energy_demand_kWh(
+            time_slot, self._calc_energy_to_buy_maximum(time_slot))
+
+    def _calc_Q_consumption(self, time_slot: DateTime, consumption_kWh: float) -> float:
```

Comment:
From Google: `Q is the net heat transferred into/out of the system. And the unit for heat is Joules.`
Since you're using this function for both `_calc_temp_decrease_K` and `_calc_temp_increase_K`, I would suggest to rename it to something else to better fit the usage (e.g. `energy_kWh_to_heat_joules`). If you agree this, it would be nice to also rename `consumption_kWh` to `energy_kWh`.

Reply:
`1 J` can also be rewritten as` 1 Ws` ([wikipedia](https://de.wikipedia.org/wiki/Joule)) which in turn can be converted into `kWh` which is widely used in our codebase. With `Q` I was following the specs on the [requirement document](https://docs.google.com/document/d/1r_oZh6dFpz5iHdO3MI4w4JpDr_WKc0L3PnUprHLiQw4/edit#).
I am open to rename the `_Q_` into `_energy_` or '_heat_' in the whole module, but I vote for keeping `kWh` as unit. What do you think?

Reply:
I am also in your side to stick with kWh and not introduce any new concepts (heat/joules) in the code. What I say is that the HeatPump is itself very convoluted and our code should not add to its complexity.
Since the `_calc_Q_consumption` is used for both heat consumed and produced, how about we remove the trailing `_consumption` from its name? Something like:
`_calc_Q_from_energy_kWh` or `_calc_Q_from_kWh`?

Reply:
Good idea! IMO `_calc_Q_from_energy_kWh` is most fitting. let me rename it.

Reply:
How about this: [a6c376d](https://github.com/gridsingularity/gsy-e/pull/1581/commits/a6c376dc1914bf35371c1e5a12a9d94fda9711a4)

Reply:
I like it very much now! Thank you.